File: assets/backing-services/typescript/tracing.ts
/**
 * This process's traces: the SDK, wired; the exporter, only when somewhere was named to send to.
 *
 * ── Why the SDK ships and the exporter does not ───────────────────────────────────────────────────────
 * A trace id is worth having before anything collects it. It is what ties a log line to the request that
 * produced it, what an incoming `traceparent` carries in from whoever called this service, and — for a
 * project that records events — what a business transaction is correlated by. None of that needs a
 * collector, and all of it needs the SDK.
 *
 * What a collector *would* need is an address, and there is no honest default for one: a starter that
 * points at `http://localhost:4318` either finds nothing there, and then spends the first minute of every
 * run retrying a connection nobody asked for, or finds something and starts shipping a project's traffic
 * somewhere it was never told about. So the rule is the one the endpoint states: set
 * `OTEL_EXPORTER_OTLP_ENDPOINT` and spans are exported; leave it unset and they are recorded, given ids,
 * and dropped. The service starts and `make verify` passes with nothing listening either way.
 *
 * ── Read through the checked environment, not `process.env` ───────────────────────────────────────────
 * The endpoint arrives as `app.config`, which `src/config.ts` declares and `@fastify/env` has already
 * validated — so a value that is not a URL stops the process at start-up with the variable named, rather
 * than being discovered as an exporter that quietly never connects. That is also why `main.ts` starts
 * tracing *after* `app.ready()` and *before* `app.listen()`: `ready()` is where the environment is checked
 * and the instrumentation plugin is loaded, `listen()` is where the first request could arrive, and the
 * window between them is the one moment both facts are true.
 *
 * Registering the provider that late is safe because `@opentelemetry/api` hands out a proxy tracer: the
 * instrumentation asks for its tracer while the app boots, and that tracer starts delegating to this
 * provider the moment `register()` is called.
 *
 * ── An unreachable collector is a warning, never a crash ──────────────────────────────────────────────
 * Exporting happens on a background batch, off the request path, and the exporter reports a failure
 * through OpenTelemetry's own `diag` channel. That channel is wired to this service's logger below, so an
 * endpoint pointing at nothing produces warnings and a service that still answers. Telemetry that can take
 * the service down with it is worse than no telemetry.
 */
import { DiagLogLevel, diag, trace } from '@opentelemetry/api';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { BatchSpanProcessor, NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } from '@opentelemetry/semantic-conventions';

/** Just the two values tracing reads, so a test hands over an object rather than a whole `Config`. */
export type TracingConfig = {
  OTEL_SERVICE_NAME: string;
  /** Empty is "no exporter", which is the same answer as unset — see `src/config.ts`. */
  OTEL_EXPORTER_OTLP_ENDPOINT: string;
};

/** What `startTracing` hands back: whether anything is being shipped, and how to stop. */
export type Tracing = {
  /** `true` only when an endpoint was named — the one fact a start-up line should report. */
  readonly exporting: boolean;
  /** Flush what is batched and release the exporter. Called from the entry point's signal handler. */
  shutdown(): Promise<void>;
};

/** How a warning from inside OpenTelemetry reaches this process's log. */
export type TracingLogger = { warn(message: string): void; error(message: string): void };

/**
 * Start recording spans, and export them only where `OTEL_EXPORTER_OTLP_ENDPOINT` says to.
 *
 * `register()` is what makes the provider global: it installs the async-hooks context manager, so
 * `trace.getActiveSpan()` answers inside a handler, and the W3C propagator, so an incoming `traceparent`
 * continues the caller's trace rather than starting a new one.
 *
 * @param config the checked environment — `app.config`, or a plain object in a test
 * @param logger where OpenTelemetry's own diagnostics go; the app's logger in the entry point
 */
export function startTracing(config: TracingConfig, logger: TracingLogger): Tracing {
  diag.setLogger(
    {
      error: (m) => logger.error(m),
      warn: (m) => logger.warn(m),
      info: () => {},
      debug: () => {},
      verbose: () => {},
    },
    DiagLogLevel.WARN,
  );
  const endpoint =
    config.OTEL_EXPORTER_OTLP_ENDPOINT === '' ? undefined : config.OTEL_EXPORTER_OTLP_ENDPOINT;
  const provider = new NodeTracerProvider({
    resource: resourceFromAttributes({
      [ATTR_SERVICE_NAME]: config.OTEL_SERVICE_NAME,
      [ATTR_SERVICE_VERSION]: '0.1.0',
    }),
    // No processor at all is the "no exporter" state, and it is not the same as a disabled SDK: spans are
    // still created, so every id below is real and every log line still carries one. They are simply not
    // kept once they end.
    spanProcessors:
      endpoint === undefined
        ? []
        : [
            new BatchSpanProcessor(
              new OTLPTraceExporter({ url: `${endpoint.replace(/\/$/, '')}/v1/traces` }),
            ),
          ],
  });
  provider.register();
  return {
    exporting: endpoint !== undefined,
    // Flushing is a network call, and the last thing a process does before exiting is the worst moment
    // for one to be unhandled: a collector that has gone away would otherwise turn a clean shutdown into
    // a non-zero exit and an orchestrator's restart loop. The failure is reported and swallowed, here
    // rather than at the call site, so nobody has to remember to.
    shutdown: async () => {
      try {
        await provider.shutdown();
      } catch (error) {
        logger.warn(`traces could not be flushed on shutdown: ${String(error)}`);
      }
    },
  };
}

/**
 * The trace and span currently in scope, as the ids an event carries — or nothing outside a span.
 *
 * ── Why this exists ───────────────────────────────────────────────────────────────────────────────────
 * Where this project records events, the port has required a `correlationId` and an optional
 * `causationId` on every one of them from the first line — and until there was a trace to take them from,
 * every slice had to invent both. Inventing them is how a causal tree ends up with everything appearing to have caused
 * itself. The request already has an identity — the span the transport opened for it, continuing whatever
 * `traceparent` the caller sent — so that is what the events it produces are correlated by, and a trace
 * that crossed two services correlates the events on both sides of it.
 *
 * ── Why they are re-punctuated rather than re-encoded ─────────────────────────────────────────────────
 * A correlation id is a UUID, and a W3C trace id is the same 128 bits written without hyphens: putting
 * them back is a formatting change and nothing is invented. A span id is 64 bits, half of a UUID, so it
 * is placed in the low half with the high half left as zeroes — reversible, and the zero prefix is what
 * says at a glance that this id came from a span rather than from `randomUUID()`.
 *
 * ── How a slice uses it ───────────────────────────────────────────────────────────────────────────────
 * ```ts
 * const { correlationId, causationId } = traceIds();
 * const event: DomainEvent = {
 *   type: 'OrderPlaced', schemaVersion: 1, streamId, payload, occurredAt, actor,
 *   correlationId: createCorrelationId(correlationId ?? randomUUID()),
 *   ...(causationId === undefined ? {} : { causationId: createCausationId(causationId) }),
 * };
 * ```
 * `createCorrelationId` still parses them, because the branded types exist to be checked once at the edge
 * and this is an edge like any other.
 */
export function traceIds(): { correlationId?: string; causationId?: string } {
  const span = trace.getActiveSpan();
  if (span === undefined) {
    return {};
  }
  const { traceId, spanId } = span.spanContext();
  return { correlationId: hyphenate(traceId), causationId: hyphenate(`0000000000000000${spanId}`) };
}

/** 32 hex characters as a UUID reads them: 8-4-4-4-12. */
function hyphenate(hex: string): string {
  return [
    hex.slice(0, 8),
    hex.slice(8, 12),
    hex.slice(12, 16),
    hex.slice(16, 20),
    hex.slice(20, 32),
  ].join('-');
}

/**
 * What every log record gains: the trace and span it happened inside, or nothing outside a request.
 *
 * pino calls this per record, which is the only way it can work — the logger is built before the app and
 * the span exists only while a request is being served. `trace_id` and `span_id` are the spellings
 * OpenTelemetry's own logging conventions use, so a collector correlates a log line with its trace without
 * being told how.
 */
export function traceContextMixin(): Record<string, string> {
  const span = trace.getActiveSpan();
  if (span === undefined) {
    return {};
  }
  const { traceId, spanId } = span.spanContext();
  return { trace_id: traceId, span_id: spanId };
}

File: assets/backing-services/python/tracing.py
"""This process's traces: the SDK, wired; the exporter, only when somewhere was named to send to.

── Why the SDK ships and the exporter does not ──────────────────────────────────────────────
A trace id is worth having before anything collects it. It is what ties a log line to the
request that produced it, what an incoming ``traceparent`` carries in from whoever called this
service, and — for a project that records events — what a business transaction is correlated
by. None of that needs a collector, and all of it needs the SDK.

What a collector *would* need is an address, and there is no honest default for one. A starter
pointing at ``http://localhost:4318`` either finds nothing there and retries a connection
nobody asked for, or finds something and ships a project's traffic somewhere it was never told
about. So the rule is the one the variable states: set ``OTEL_EXPORTER_OTLP_ENDPOINT`` and
spans are exported; leave it unset and they are recorded, given ids, and dropped. The service
starts and ``make verify`` passes with nothing listening either way.

── Read through the checked environment, not ``os.environ`` ─────────────────────────────────
The endpoint arrives as the ``Settings`` model ``settings.py`` declares and ``main`` has
already validated, so a value that is not a URL stops the process at start-up with the variable
named rather than being discovered as an exporter that silently never connects.

── An unreachable collector is a warning, never a crash ─────────────────────────────────────
Exporting happens on a background batch, off the request path, and a failure is reported
through OpenTelemetry's own logger — which is this process's logger, because
``logging_setup`` configured the root. Telemetry that can take the service down with it is
worse than no telemetry.
"""

from __future__ import annotations

import logging
import uuid
from dataclasses import dataclass

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import SERVICE_NAME, SERVICE_VERSION, Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor


@dataclass(frozen=True)
class Tracing:
    """What ``start_tracing`` hands back: whether anything is shipped, and how to stop."""

    #: ``True`` only when an endpoint was named — the one fact a start-up line should report.
    exporting: bool
    provider: TracerProvider

    def shutdown(self) -> None:
        """Flush what is batched and release the exporter.

        A flush is a network call, and process exit is the worst moment for one to raise: a
        collector that has gone away would otherwise turn a clean shutdown into a traceback.
        Reported and swallowed here rather than at the call site, so nobody has to remember to.
        """
        try:
            self.provider.shutdown()
        except Exception as error:
            logging.getLogger(__name__).warning(
                "traces could not be flushed on shutdown: %s", error
            )


def start_tracing(service_name: str, endpoint: str | None) -> Tracing:
    """Record spans, and export them only where ``OTEL_EXPORTER_OTLP_ENDPOINT`` says to.

    ``set_tracer_provider`` is what makes this the process's provider, so anything already
    holding a tracer — the FastAPI instrumentation, which took one while the app was built —
    starts recording through it.

    :param service_name: what this service calls itself in a trace
    :param endpoint: the OTLP collector's base URL, or ``None`` for no exporter at all
    """
    resource = Resource.create({SERVICE_NAME: service_name, SERVICE_VERSION: "0.1.0"})
    provider = TracerProvider(resource=resource)
    if endpoint is not None:
        # No processor at all is the "no exporter" state, and it is not the same as a disabled
        # SDK: spans are still created, so every id below is real and every log line still
        # carries one. They are simply not kept once they end.
        exporter = OTLPSpanExporter(endpoint=f"{endpoint.rstrip('/')}/v1/traces")
        provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)
    return Tracing(exporting=endpoint is not None, provider=provider)


def trace_ids() -> tuple[uuid.UUID | None, uuid.UUID | None]:
    """The trace and span in scope, as the ids an event carries — ``(None, None)`` outside one.

    ── Why this exists ──────────────────────────────────────────────────────────────────────
    Where this project records events, the port has required a ``correlation_id`` and an
    optional ``causation_id`` on every one of them from the first line — and until there was
    a trace to take them from, every slice had to invent both. Inventing them is how a causal
    tree ends up with everything appearing to have caused itself. The request already has an
    identity — the span the transport opened for it, continuing whatever ``traceparent`` the
    caller sent — so that is what the events it produces are correlated by, and a trace that
    crossed two services correlates the events on both sides of it.

    ── Why they are re-punctuated rather than re-encoded ────────────────────────────────────
    A correlation id is a UUID, and a W3C trace id is the same 128 bits: reading one as the
    other invents nothing. A span id is 64 bits, half of a UUID, so it goes in the low half
    with the high half left zero — reversible, and the zero prefix is what says at a glance
    that the id came from a span rather than from ``uuid4()``.

    ── How a slice uses it ──────────────────────────────────────────────────────────────────
    .. code-block:: python

        correlation, causation = trace_ids()
        event = DomainEvent(
            ...,
            correlation_id=create_correlation_id(correlation or uuid.uuid4()),
            causation_id=None if causation is None else create_causation_id(causation),
        )
    """
    span = trace.get_current_span()
    context = span.get_span_context()
    if not context.is_valid:
        return None, None
    return uuid.UUID(int=context.trace_id), uuid.UUID(int=context.span_id)


def trace_context() -> dict[str, str]:
    """``trace_id`` and ``span_id`` for a log record, or nothing outside a request.

    The spellings are OpenTelemetry's own logging conventions, so a collector correlates a log
    line with its trace without being told how. ``logging_setup`` puts them on every record.
    """
    context = trace.get_current_span().get_span_context()
    if not context.is_valid:
        return {}
    return {
        "trace_id": format(context.trace_id, "032x"),
        "span_id": format(context.span_id, "016x"),
    }
